File: ideviewer/detectors/jetbrains.py
"""
JetBrains IDE detector.

Supports: IntelliJ IDEA, PyCharm, WebStorm, GoLand, CLion, Rider, PhpStorm, RubyMine, DataGrip, Android Studio
"""

import os
import re
import sys
import xml.etree.ElementTree as ET
from pathlib import Path
from typing import List, Optional, Dict, Any, Tuple
from datetime import datetime

# Handle both direct execution and package import
try:
    from .base import BaseDetector
    from ..models import IDE, IDEType, Extension, Permission
except ImportError:
    sys.path.insert(0, str(Path(__file__).parent.parent.parent))
    from ideviewer.detectors.base import BaseDetector
    from ideviewer.models import IDE, IDEType, Extension, Permission


class JetBrainsDetector(BaseDetector):
    """Detector for JetBrains IDEs."""
    
    # JetBrains IDE configurations
    JETBRAINS_IDES = {
        IDEType.JETBRAINS_IDEA: {
            "name": "IntelliJ IDEA",
            "folder_patterns": ["IntelliJIdea*", "IdeaIC*"],
            "app_names": {
                "darwin": ["IntelliJ IDEA.app", "IntelliJ IDEA CE.app", "IntelliJ IDEA Ultimate.app"],
                "windows": ["idea64.exe", "idea.exe"],
                "linux": ["idea.sh", "idea"],
            },
            "toolbox_id": "intellij-idea",
        },
        IDEType.JETBRAINS_PYCHARM: {
            "name": "PyCharm",
            "folder_patterns": ["PyCharm*"],
            "app_names": {
                "darwin": ["PyCharm.app", "PyCharm CE.app", "PyCharm Professional.app"],
                "windows": ["pycharm64.exe", "pycharm.exe"],
                "linux": ["pycharm.sh", "pycharm"],
            },
            "toolbox_id": "pycharm",
        },
        IDEType.JETBRAINS_WEBSTORM: {
            "name": "WebStorm",
            "folder_patterns": ["WebStorm*"],
            "app_names": {
                "darwin": ["WebStorm.app"],
                "windows": ["webstorm64.exe", "webstorm.exe"],
                "linux": ["webstorm.sh", "webstorm"],
            },
            "toolbox_id": "webstorm",
        },
        IDEType.JETBRAINS_GOLAND: {
            "name": "GoLand",
            "folder_patterns": ["GoLand*"],
            "app_names": {
                "darwin": ["GoLand.app"],
                "windows": ["goland64.exe", "goland.exe"],
                "linux": ["goland.sh", "goland"],
            },
            "toolbox_id": "goland",
        },
        IDEType.JETBRAINS_CLION: {
            "name": "CLion",
            "folder_patterns": ["CLion*"],
            "app_names": {
                "darwin": ["CLion.app"],
                "windows": ["clion64.exe", "clion.exe"],
                "linux": ["clion.sh", "clion"],
            },
            "toolbox_id": "clion",
        },
        IDEType.JETBRAINS_RIDER: {
            "name": "Rider",
            "folder_patterns": ["Rider*"],
            "app_names": {
                "darwin": ["Rider.app"],
                "windows": ["rider64.exe", "rider.exe"],
                "linux": ["rider.sh", "rider"],
            },
            "toolbox_id": "rider",
        },
        IDEType.JETBRAINS_PHPSTORM: {
            "name": "PhpStorm",
            "folder_patterns": ["PhpStorm*"],
            "app_names": {
                "darwin": ["PhpStorm.app"],
                "windows": ["phpstorm64.exe", "phpstorm.exe"],
                "linux": ["phpstorm.sh", "phpstorm"],
            },
            "toolbox_id": "phpstorm",
        },
        IDEType.JETBRAINS_RUBYMINE: {
            "name": "RubyMine",
            "folder_patterns": ["RubyMine*"],
            "app_names": {
                "darwin": ["RubyMine.app"],
                "windows": ["rubymine64.exe", "rubymine.exe"],
                "linux": ["rubymine.sh", "rubymine"],
            },
            "toolbox_id": "rubymine",
        },
        IDEType.JETBRAINS_DATAGRIP: {
            "name": "DataGrip",
            "folder_patterns": ["DataGrip*"],
            "app_names": {
                "darwin": ["DataGrip.app"],
                "windows": ["datagrip64.exe", "datagrip.exe"],
                "linux": ["datagrip.sh", "datagrip"],
            },
            "toolbox_id": "datagrip",
        },
        IDEType.ANDROID_STUDIO: {
            "name": "Android Studio",
            "folder_patterns": ["AndroidStudio*", "Google/AndroidStudio*"],
            "app_names": {
                "darwin": ["Android Studio.app"],
                "windows": ["studio64.exe", "studio.exe"],
                "linux": ["studio.sh", "android-studio"],
            },
            "toolbox_id": "android-studio",
        },
    }
    
    def detect(self) -> List[IDE]:
        """Detect all installed JetBrains IDEs."""
        ides = []
        
        for ide_type, config in self.JETBRAINS_IDES.items():
            detected = self._detect_jetbrains_ide(ide_type, config)
            for ide in detected:
                ide.extensions = self.parse_extensions(ide)
                ides.append(ide)
        
        return ides
    
    def _get_config_base_paths(self) -> List[Path]:
        """Get base paths for JetBrains configuration."""
        paths = []
        
        if self.is_macos:
            paths.append(self.home / "Library" / "Application Support" / "JetBrains")
            paths.append(self.home / "Library" / "Preferences")
        elif self.is_linux:
            paths.append(self.home / ".config" / "JetBrains")
            paths.append(self.home / ".local" / "share" / "JetBrains")
        elif self.is_windows:
            appdata = os.environ.get("APPDATA", "")
            if appdata:
                paths.append(Path(appdata) / "JetBrains")
        
        return [p for p in paths if p.exists()]
    
    def _get_plugins_paths(self) -> List[Path]:
        """Get paths where JetBrains plugins are stored."""
        paths = []
        
        if self.is_macos:
            paths.append(self.home / "Library" / "Application Support" / "JetBrains")
        elif self.is_linux:
            paths.append(self.home / ".local" / "share" / "JetBrains")
        elif self.is_windows:
            appdata = os.environ.get("APPDATA", "")
            if appdata:
                paths.append(Path(appdata) / "JetBrains")
        
        return [p for p in paths if p.exists()]
    
    def _get_install_paths(self) -> List[Path]:
        """Get paths where JetBrains IDEs might be installed."""
        paths = []
        
        if self.is_macos:
            paths.append(Path("/Applications"))
            paths.append(self.home / "Applications")
            # JetBrains Toolbox
            paths.append(self.home / "Library" / "Application Support" / "JetBrains" / "Toolbox" / "apps")
        elif self.is_linux:
            paths.append(Path("/opt"))
            paths.append(Path("/usr/share"))
            paths.append(self.home / ".local" / "share" / "JetBrains" / "Toolbox" / "apps")
            paths.append(Path("/snap"))
        elif self.is_windows:
            program_files = os.environ.get("ProgramFiles", "C:\\Program Files")
            program_files_x86 = os.environ.get("ProgramFiles(x86)", "C:\\Program Files (x86)")
            localappdata = os.environ.get("LOCALAPPDATA", "")
            
            paths.append(Path(program_files) / "JetBrains")
            paths.append(Path(program_files_x86) / "JetBrains")
            if localappdata:
                paths.append(Path(localappdata) / "JetBrains" / "Toolbox" / "apps")
        
        return [p for p in paths if p.exists()]
    
    def _detect_jetbrains_ide(self, ide_type: IDEType, config: Dict[str, Any]) -> List[IDE]:
        """Detect a specific JetBrains IDE."""
        found_ides = []
        
        # Search for installed applications
        for install_path in self._get_install_paths():
            for app_name in config["app_names"].get(self.system, []):
                if self.is_macos:
                    app_path = install_path / app_name
                    if app_path.exists():
                        ide = self._create_ide_from_app(ide_type, config, app_path)
                        if ide:
                            found_ides.append(ide)
                else:
                    # Search recursively for the executable
                    for root, dirs, files in os.walk(install_path):
                        if app_name in files:
                            ide = self._create_ide_from_app(ide_type, config, Path(root) / app_name)
                            if ide:
                                found_ides.append(ide)
        
        # Look for config directories even if app not found in standard locations
        for config_base in self._get_config_base_paths():
            for pattern in config["folder_patterns"]:
                for config_dir in config_base.glob(pattern):
                    if config_dir.is_dir():
                        # Check if this IDE config exists but app wasn't found
                        existing = any(ide.config_path == str(config_dir) for ide in found_ides)
                        if not existing:
                            version = self._extract_version_from_folder(config_dir.name)
                            plugins_path = self._find_plugins_path(config_dir)
                            
                            ide = IDE(
                                ide_type=ide_type,
                                name=config["name"],
                                version=version,
                                install_path=None,
                                config_path=str(config_dir),
                                extensions_path=str(plugins_path) if plugins_path else None,
                            )
                            found_ides.append(ide)
        
        return found_ides
    
    def _create_ide_from_app(self, ide_type: IDEType, config: Dict[str, Any], app_path: Path) -> Optional[IDE]:
        """Create an IDE object from an application path."""
        version = None
        config_path = None
        extensions_path = None
        
        if self.is_macos and app_path.suffix == ".app":
            # Try to get version from Info.plist
            plist_path = app_path / "Contents" / "Info.plist"
            if plist_path.exists():
                version = self._get_version_from_plist(plist_path)
        
        # Find corresponding config directory
        for config_base in self._get_config_base_paths():
            for pattern in config["folder_patterns"]:
                for config_dir in config_base.glob(pattern):
                    if config_dir.is_dir():
                        config_path = str(config_dir)
                        plugins_path = self._find_plugins_path(config_dir)
                        if plugins_path:
                            extensions_path = str(plugins_path)
                        break
                if config_path:
                    break
        
        return IDE(
            ide_type=ide_type,
            name=config["name"],
            version=version,
            install_path=str(app_path),
            config_path=config_path,
            extensions_path=extensions_path,
        )
    
    def _find_plugins_path(self, config_dir: Path) -> Optional[Path]:
        """Find the plugins directory for a JetBrains config."""
        # Try common plugin locations
        candidates = [
            config_dir / "plugins",
            config_dir.parent / (config_dir.name + "-plugins"),
        ]
        
        for candidate in candidates:
            if candidate.exists():
                return candidate
        
        return None
    
    def _extract_version_from_folder(self, folder_name: str) -> Optional[str]:
        """Extract version number from folder name like 'PyCharm2023.2'."""
        match = re.search(r'(\d+\.\d+(?:\.\d+)?)', folder_name)
        return match.group(1) if match else None
    
    def _get_version_from_plist(self, plist_path: Path) -> Optional[str]:
        """Get version from macOS Info.plist."""
        try:
            import plistlib
            with open(plist_path, "rb") as f:
                plist = plistlib.load(f)
                return plist.get("CFBundleShortVersionString") or plist.get("CFBundleVersion")
        except Exception:
            return None
    
    def parse_extensions(self, ide: IDE) -> List[Extension]:
        """Parse plugins for a JetBrains IDE."""
        extensions = []
        
        if not ide.extensions_path or not os.path.isdir(ide.extensions_path):
            return extensions
        
        plugins_dir = Path(ide.extensions_path)
        
        for plugin_folder in plugins_dir.iterdir():
            if not plugin_folder.is_dir():
                continue
            
            if plugin_folder.name.startswith("."):
                continue
            
            extension = self._parse_plugin(plugin_folder)
            if extension:
                extensions.append(extension)
        
        return extensions
    
    def _parse_plugin(self, plugin_folder: Path) -> Optional[Extension]:
        """Parse a JetBrains plugin from its folder."""
        # JetBrains plugins can have plugin.xml in different locations
        plugin_xml_paths = [
            plugin_folder / "META-INF" / "plugin.xml",
            plugin_folder / "lib" / "plugin.xml",
        ]
        
        # Also check inside JAR files in lib/
        lib_dir = plugin_folder / "lib"
        if lib_dir.exists():
            for jar_file in lib_dir.glob("*.jar"):
                # For now, we'll just use folder-based detection
                pass
        
        plugin_xml = None
        for path in plugin_xml_paths:
            if path.exists():
                plugin_xml = path
                break
        
        if plugin_xml:
            return self._parse_plugin_xml(plugin_xml, plugin_folder)
        else:
            # Create basic extension from folder name
            return Extension(
                id=plugin_folder.name,
                name=plugin_folder.name,
                version="unknown",
                install_path=str(plugin_folder),
            )
    
    def _parse_plugin_xml(self, plugin_xml: Path, plugin_folder: Path) -> Optional[Extension]:
        """Parse plugin.xml file."""
        try:
            tree = ET.parse(plugin_xml)
            root = tree.getroot()
        except ET.ParseError:
            return None
        
        # Extract info from XML
        plugin_id = root.findtext("id", plugin_folder.name)
        name = root.findtext("name", plugin_folder.name)
        version = root.findtext("version", "unknown")
        
        # Vendor info
        vendor_elem = root.find("vendor")
        publisher = None
        homepage = None
        if vendor_elem is not None:
            publisher = vendor_elem.text
            homepage = vendor_elem.get("url")
        
        description = root.findtext("description", "")
        
        # Parse permissions/extensions
        permissions = self._extract_jetbrains_permissions(root)
        
        # Get dependencies
        dependencies = []
        for dep in root.findall(".//depends"):
            if dep.text:
                dependencies.append(dep.text)
        
        # Last updated
        last_updated = None
        try:
            stat = plugin_xml.stat()
            last_updated = datetime.fromtimestamp(stat.st_mtime)
        except OSError:
            pass
        
        return Extension(
            id=plugin_id,
            name=name,
            version=version,
            publisher=publisher,
            homepage=homepage,
            install_path=str(plugin_folder),
            permissions=permissions,
            dependencies=dependencies,
            last_updated=last_updated,
            description=description[:200] if description else None,  # Truncate
        )
    
    def _extract_jetbrains_permissions(self, root: ET.Element) -> List[Permission]:
        """Extract permissions from JetBrains plugin XML."""
        permissions = []
        
        # Check for extension points that indicate capabilities
        for ext in root.findall(".//extensions"):
            for child in ext:
                tag = child.tag
                
                # Map extension points to permissions
                if "action" in tag.lower():
                    permissions.append(Permission(
                        name="actions",
                        description="Registers IDE actions",
                        is_dangerous=False,
                    ))
                elif "toolwindow" in tag.lower():
                    permissions.append(Permission(
                        name="toolWindow",
                        description="Creates tool windows",
                        is_dangerous=False,
                    ))
                elif "projectservice" in tag.lower() or "applicationservice" in tag.lower():
                    permissions.append(Permission(
                        name="services",
                        description="Registers application/project services",
                        is_dangerous=False,
                    ))
        
        # Check for dangerous capabilities
        for ext in root.findall(".//*"):
            text = ET.tostring(ext, encoding='unicode', method='text').lower()
            
            if "exec" in text or "process" in text:
                permissions.append(Permission(
                    name="processExecution",
                    description="May execute external processes",
                    is_dangerous=True,
                ))
                break
        
        return permissions

File: pkg/detectors/jetbrains.go
package detectors

import (
	"encoding/xml"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"github.com/securient/ideviewer-oss/pkg/scanner"
)

type jetbrainsIDEConfig struct {
	ideType        scanner.IDEType
	name           string
	folderPatterns []string
	appNames       map[string][]string
	toolboxID      string
}

var jetbrainsIDEs = []jetbrainsIDEConfig{
	{
		ideType:        scanner.IDETypeJetBrainsIDEA,
		name:           "IntelliJ IDEA",
		folderPatterns: []string{"IntelliJIdea*", "IdeaIC*"},
		appNames: map[string][]string{
			"darwin":  {"IntelliJ IDEA.app", "IntelliJ IDEA CE.app", "IntelliJ IDEA Ultimate.app"},
			"windows": {"idea64.exe", "idea.exe"},
			"linux":   {"idea.sh", "idea"},
		},
		toolboxID: "intellij-idea",
	},
	{
		ideType:        scanner.IDETypeJetBrainsPyCharm,
		name:           "PyCharm",
		folderPatterns: []string{"PyCharm*"},
		appNames: map[string][]string{
			"darwin":  {"PyCharm.app", "PyCharm CE.app", "PyCharm Professional.app"},
			"windows": {"pycharm64.exe", "pycharm.exe"},
			"linux":   {"pycharm.sh", "pycharm"},
		},
		toolboxID: "pycharm",
	},
	{
		ideType:        scanner.IDETypeJetBrainsWebStorm,
		name:           "WebStorm",
		folderPatterns: []string{"WebStorm*"},
		appNames: map[string][]string{
			"darwin":  {"WebStorm.app"},
			"windows": {"webstorm64.exe", "webstorm.exe"},
			"linux":   {"webstorm.sh", "webstorm"},
		},
		toolboxID: "webstorm",
	},
	{
		ideType:        scanner.IDETypeJetBrainsGoLand,
		name:           "GoLand",
		folderPatterns: []string{"GoLand*"},
		appNames: map[string][]string{
			"darwin":  {"GoLand.app"},
			"windows": {"goland64.exe", "goland.exe"},
			"linux":   {"goland.sh", "goland"},
		},
		toolboxID: "goland",
	},
	{
		ideType:        scanner.IDETypeJetBrainsCLion,
		name:           "CLion",
		folderPatterns: []string{"CLion*"},
		appNames: map[string][]string{
			"darwin":  {"CLion.app"},
			"windows": {"clion64.exe", "clion.exe"},
			"linux":   {"clion.sh", "clion"},
		},
		toolboxID: "clion",
	},
	{
		ideType:        scanner.IDETypeJetBrainsRider,
		name:           "Rider",
		folderPatterns: []string{"Rider*"},
		appNames: map[string][]string{
			"darwin":  {"Rider.app"},
			"windows": {"rider64.exe", "rider.exe"},
			"linux":   {"rider.sh", "rider"},
		},
		toolboxID: "rider",
	},
	{
		ideType:        scanner.IDETypeJetBrainsPhpStorm,
		name:           "PhpStorm",
		folderPatterns: []string{"PhpStorm*"},
		appNames: map[string][]string{
			"darwin":  {"PhpStorm.app"},
			"windows": {"phpstorm64.exe", "phpstorm.exe"},
			"linux":   {"phpstorm.sh", "phpstorm"},
		},
		toolboxID: "phpstorm",
	},
	{
		ideType:        scanner.IDETypeJetBrainsRubyMine,
		name:           "RubyMine",
		folderPatterns: []string{"RubyMine*"},
		appNames: map[string][]string{
			"darwin":  {"RubyMine.app"},
			"windows": {"rubymine64.exe", "rubymine.exe"},
			"linux":   {"rubymine.sh", "rubymine"},
		},
		toolboxID: "rubymine",
	},
	{
		ideType:        scanner.IDETypeJetBrainsDataGrip,
		name:           "DataGrip",
		folderPatterns: []string{"DataGrip*"},
		appNames: map[string][]string{
			"darwin":  {"DataGrip.app"},
			"windows": {"datagrip64.exe", "datagrip.exe"},
			"linux":   {"datagrip.sh", "datagrip"},
		},
		toolboxID: "datagrip",
	},
	{
		ideType:        scanner.IDETypeAndroidStudio,
		name:           "Android Studio",
		folderPatterns: []string{"AndroidStudio*", "Google/AndroidStudio*"},
		appNames: map[string][]string{
			"darwin":  {"Android Studio.app"},
			"windows": {"studio64.exe", "studio.exe"},
			"linux":   {"studio.sh", "android-studio"},
		},
		toolboxID: "android-studio",
	},
}

// JetBrainsDetector detects JetBrains IDE installations.
type JetBrainsDetector struct{}

func (d *JetBrainsDetector) Name() string { return "jetbrains" }

func (d *JetBrainsDetector) Detect() ([]scanner.IDE, error) {
	var ides []scanner.IDE
	for _, cfg := range jetbrainsIDEs {
		found := detectJetBrainsIDE(cfg)
		for i := range found {
			found[i].Extensions = parseJetBrainsPlugins(found[i].ExtensionsPath)
		}
		ides = append(ides, found...)
	}
	return ides, nil
}

func configBasePaths() []string {
	home := HomeDir()
	plat := PlatformKey()
	var paths []string
	switch plat {
	case "darwin":
		paths = append(paths,
			filepath.Join(home, "Library", "Application Support", "JetBrains"),
			filepath.Join(home, "Library", "Preferences"),
		)
	case "linux":
		paths = append(paths,
			filepath.Join(home, ".config", "JetBrains"),
			filepath.Join(home, ".local", "share", "JetBrains"),
		)
	case "windows":
		if appdata := os.Getenv("APPDATA"); appdata != "" {
			paths = append(paths, filepath.Join(appdata, "JetBrains"))
		}
	}
	var existing []string
	for _, p := range paths {
		if PathExists(p) {
			existing = append(existing, p)
		}
	}
	return existing
}

func installPaths() []string {
	home := HomeDir()
	plat := PlatformKey()
	var paths []string
	switch plat {
	case "darwin":
		paths = append(paths,
			"/Applications",
			filepath.Join(home, "Applications"),
			filepath.Join(home, "Library", "Application Support", "JetBrains", "Toolbox", "apps"),
		)
	case "linux":
		paths = append(paths,
			"/opt",
			"/usr/share",
			filepath.Join(home, ".local", "share", "JetBrains", "Toolbox", "apps"),
			"/snap",
		)
	case "windows":
		pf := os.Getenv("ProgramFiles")
		if pf == "" {
			pf = `C:\Program Files`
		}
		pf86 := os.Getenv("ProgramFiles(x86)")
		if pf86 == "" {
			pf86 = `C:\Program Files (x86)`
		}
		paths = append(paths,
			filepath.Join(pf, "JetBrains"),
			filepath.Join(pf86, "JetBrains"),
		)
		if localappdata := os.Getenv("LOCALAPPDATA"); localappdata != "" {
			paths = append(paths, filepath.Join(localappdata, "JetBrains", "Toolbox", "apps"))
		}
	}
	var existing []string
	for _, p := range paths {
		if PathExists(p) {
			existing = append(existing, p)
		}
	}
	return existing
}

func detectJetBrainsIDE(cfg jetbrainsIDEConfig) []scanner.IDE {
	plat := PlatformKey()
	var found []scanner.IDE

	// Search installed applications
	for _, ip := range installPaths() {
		for _, appName := range cfg.appNames[plat] {
			if plat == "darwin" {
				appPath := filepath.Join(ip, appName)
				if PathExists(appPath) {
					ide := createJetBrainsIDE(cfg, appPath)
					found = append(found, ide)
				}
			} else {
				// Walk to find executables
				_ = filepath.WalkDir(ip, func(path string, d os.DirEntry, err error) error {
					if err != nil {
						return filepath.SkipDir
					}
					if d.Name() == appName && !d.IsDir() {
						ide := createJetBrainsIDE(cfg, path)
						found = append(found, ide)
					}
					return nil
				})
			}
		}
	}

	// Look for config directories even if app not found
	for _, base := range configBasePaths() {
		for _, pattern := range cfg.folderPatterns {
			matches, _ := filepath.Glob(filepath.Join(base, pattern))
			for _, m := range matches {
				info, err := os.Stat(m)
				if err != nil || !info.IsDir() {
					continue
				}
				// Skip if we already found this config
				alreadyFound := false
				for _, f := range found {
					if f.ConfigPath == m {
						alreadyFound = true
						break
					}
				}
				if alreadyFound {
					continue
				}

				version := extractVersionFromFolder(filepath.Base(m))
				pluginsPath := findPluginsPath(m)
				ide := scanner.IDE{
					IDEType:        cfg.ideType,
					Name:           cfg.name,
					Version:        version,
					ConfigPath:     m,
					ExtensionsPath: pluginsPath,
				}
				found = append(found, ide)
			}
		}
	}

	return found
}

func createJetBrainsIDE(cfg jetbrainsIDEConfig, appPath string) scanner.IDE {
	var version, configPath, extensionsPath string

	// macOS: try to read version from Info.plist (just extract from config folder)
	for _, base := range configBasePaths() {
		for _, pattern := range cfg.folderPatterns {
			matches, _ := filepath.Glob(filepath.Join(base, pattern))
			for _, m := range matches {
				info, err := os.Stat(m)
				if err != nil || !info.IsDir() {
					continue
				}
				configPath = m
				if pp := findPluginsPath(m); pp != "" {
					extensionsPath = pp
				}
				if version == "" {
					version = extractVersionFromFolder(filepath.Base(m))
				}
				break
			}
			if configPath != "" {
				break
			}
		}
		if configPath != "" {
			break
		}
	}

	return scanner.IDE{
		IDEType:        cfg.ideType,
		Name:           cfg.name,
		Version:        version,
		InstallPath:    appPath,
		ConfigPath:     configPath,
		ExtensionsPath: extensionsPath,
	}
}

func findPluginsPath(configDir string) string {
	candidates := []string{
		filepath.Join(configDir, "plugins"),
		configDir + "-plugins",
	}
	for _, c := range candidates {
		if PathExists(c) {
			return c
		}
	}
	return ""
}

var versionRe = regexp.MustCompile(`(\d+\.\d+(?:\.\d+)?)`)

func extractVersionFromFolder(name string) string {
	m := versionRe.FindString(name)
	return m
}

// XML structures for plugin.xml parsing.
type pluginXML struct {
	XMLName     xml.Name        `xml:"idea-plugin"`
	ID          string          `xml:"id"`
	Name        string          `xml:"name"`
	Version     string          `xml:"version"`
	Vendor      pluginVendor    `xml:"vendor"`
	Description string          `xml:"description"`
	Depends     []string        `xml:"depends"`
	Extensions  []pluginExtElem `xml:"extensions"`
}

type pluginVendor struct {
	URL  string `xml:"url,attr"`
	Text string `xml:",chardata"`
}

type pluginExtElem struct {
	Children []xmlChild `xml:",any"`
}

type xmlChild struct {
	XMLName xml.Name
}

func parseJetBrainsPlugins(pluginsDir string) []scanner.Extension {
	if pluginsDir == "" {
		return nil
	}
	entries, err := os.ReadDir(pluginsDir)
	if err != nil {
		return nil
	}
	var exts []scanner.Extension
	for _, entry := range entries {
		if !entry.IsDir() || strings.HasPrefix(entry.Name(), ".") {
			continue
		}
		ext := parseJetBrainsPlugin(filepath.Join(pluginsDir, entry.Name()))
		exts = append(exts, ext)
	}
	return exts
}

func parseJetBrainsPlugin(dir string) scanner.Extension {
	// Try known plugin.xml locations
	candidates := []string{
		filepath.Join(dir, "META-INF", "plugin.xml"),
		filepath.Join(dir, "lib", "plugin.xml"),
	}
	for _, c := range candidates {
		if data, err := os.ReadFile(c); err == nil {
			if ext, ok := parsePluginXML(data, dir, c); ok {
				return ext
			}
		}
	}
	// Fallback: basic extension from folder name
	return scanner.Extension{
		ID:          filepath.Base(dir),
		Name:        filepath.Base(dir),
		Version:     "unknown",
		InstallPath: dir,
		Enabled:     true,
	}
}

func parsePluginXML(data []byte, dir, xmlPath string) (scanner.Extension, bool) {
	var p pluginXML
	if err := xml.Unmarshal(data, &p); err != nil {
		return scanner.Extension{}, false
	}

	id := p.ID
	if id == "" {
		id = filepath.Base(dir)
	}
	name := p.Name
	if name == "" {
		name = filepath.Base(dir)
	}
	version := p.Version
	if version == "" {
		version = "unknown"
	}

	publisher := strings.TrimSpace(p.Vendor.Text)
	homepage := strings.TrimSpace(p.Vendor.URL)

	desc := p.Description
	if len(desc) > 200 {
		desc = desc[:200]
	}

	// Permissions from extension points
	perms := extractJetBrainsPermissions(data, p)

	var lastUpdated *time.Time
	if info, err := os.Stat(xmlPath); err == nil {
		t := info.ModTime()
		lastUpdated = &t
	}

	return scanner.Extension{
		ID:           id,
		Name:         name,
		Version:      version,
		Publisher:    publisher,
		Homepage:     homepage,
		Description:  desc,
		InstallPath:  dir,
		Permissions:  perms,
		Dependencies: p.Depends,
		Enabled:      true,
		LastUpdated:  lastUpdated,
	}, true
}

func extractJetBrainsPermissions(raw []byte, p pluginXML) []scanner.Permission {
	var perms []scanner.Permission
	seen := make(map[string]bool)

	for _, ext := range p.Extensions {
		for _, child := range ext.Children {
			tag := strings.ToLower(child.XMLName.Local)
			if strings.Contains(tag, "action") && !seen["actions"] {
				seen["actions"] = true
				perms = append(perms, scanner.Permission{Name: "actions", Description: "Registers IDE actions", IsDangerous: false})
			}
			if strings.Contains(tag, "toolwindow") && !seen["toolWindow"] {
				seen["toolWindow"] = true
				perms = append(perms, scanner.Permission{Name: "toolWindow", Description: "Creates tool windows", IsDangerous: false})
			}
			if (strings.Contains(tag, "projectservice") || strings.Contains(tag, "applicationservice")) && !seen["services"] {
				seen["services"] = true
				perms = append(perms, scanner.Permission{Name: "services", Description: "Registers application/project services", IsDangerous: false})
			}
		}
	}

	// Check for dangerous patterns in the raw XML
	lower := strings.ToLower(string(raw))
	if (strings.Contains(lower, "exec") || strings.Contains(lower, "process")) && !seen["processExecution"] {
		seen["processExecution"] = true
		perms = append(perms, scanner.Permission{Name: "processExecution", Description: "May execute external processes", IsDangerous: true})
	}

	return perms
}
